### PR TITLE
deploy: make retryTimeout proportional to deployer Pod startTime

### PR DESCRIPTION
**Pull request by mfojtik**
Fixes: https://github.com/openshift/origin/issues/12154

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

@kargakis ptal

**Comment by mfojtik** [comment]
TODO: need a test

**Comment by mfojtik** [comment]
[test]

File: pkg/deploy/strategy/recreate/recreate.go [hunk 1]
```diff
@@ -110,9 +114,21 @@ func (s *RecreateDeploymentStrategy) DeployWithAcceptor(from *kapi.ReplicationCo
 		return fmt.Errorf("couldn't decode config from deployment %s: %v", to.Name, err)
 	}
 
+	deployerPod, err := s.podClient.Pods(to.Namespace).Get(deployutil.DeployerPodNameForDeployment(to.Name))
+	if err != nil {
+		return fmt.Errorf("unable to get deployer pod for %s/%s: %v", to.Namespace, to.Name, err)
```
Comment (kargakis) [inline pkg/deploy/strategy/recreate/recreate.go]: You can also fallback to the current timeout in case of this error

Comment (mfojtik) [inline pkg/deploy/strategy/recreate/recreate.go]: ok (altough we fail if you have hooks as we request deployer pod there)

**Comment by mfojtik** [comment]
@kargakis fixed.

File: pkg/deploy/strategy/recreate/recreate.go [hunk 2]
```diff
@@ -110,9 +115,23 @@ func (s *RecreateDeploymentStrategy) DeployWithAcceptor(from *kapi.ReplicationCo
 		return fmt.Errorf("couldn't decode config from deployment %s: %v", to.Name, err)
 	}
 
+	timeoutSeconds := defaultRetryTimeout
+	deployerPod, err := s.podClient.Pods(to.Namespace).Get(deployutil.DeployerPodNameForDeployment(to.Name))
+	if err != nil {
+		// if we are not able to get the deployer pod, set the timeout to default and just log
+		// the error.
+		glog.Errorf("unable to get deployer pod for %s/%s: %v", to.Namespace, to.Name, err)
```
Comment (kargakis) [inline pkg/deploy/strategy/recreate/recreate.go]: We don't really care about this error. I would suggest something like
```go
if err == nil {
  // set timeout - startTime
}
```
and be done with it

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
@kargakis test added to prove this work, PTAL

File: pkg/deploy/strategy/recreate/recreate.go [hunk 3]
```diff
@@ -110,9 +117,17 @@ func (s *RecreateDeploymentStrategy) DeployWithAcceptor(from *kapi.ReplicationCo
 		return fmt.Errorf("couldn't decode config from deployment %s: %v", to.Name, err)
 	}
 
+	retryTimeout := time.Duration(*config.Spec.Strategy.RecreateParams.TimeoutSeconds * int64(time.Second))
```
Comment (kargakis) [inline pkg/deploy/strategy/recreate/recreate.go]: Why not `time.Duration(*config.Spec.Strategy.RecreateParams.TimeoutSeconds)*time.Second`? Wrapping time.Second with int64 is ugly.

**Comment by kargakis** [comment]
LGTM

**Comment by mfojtik** [comment]
[merge]

File: pkg/deploy/api/types.go [hunk 4]
```diff
@@ -10,6 +10,8 @@ import (
 const (
 	// DefaultRollingTimeoutSeconds is the default TimeoutSeconds for RollingDeploymentStrategyParams.
 	DefaultRollingTimeoutSeconds int64 = 10 * 60
+	// DefaultRecreateTimeoutSeconds is the default TimeoutSeconds for RecreateDeploymentStrategyParams.
+	DefaultRecreateTimeoutSeconds int64 = 2 * 60
```
Comment (mfojtik) [inline pkg/deploy/api/types.go]: we should default this in defaulter for recreate?

File: pkg/deploy/strategy/recreate/recreate.go [hunk 5]
```diff
@@ -120,20 +120,42 @@ func (s *RecreateDeploymentStrategy) DeployWithAcceptor(from *kapi.ReplicationCo
...
+	fmt.Fprintf(s.out, "--> Waiting up to %s for rollout to finish\n", retryTimeout)
 
-	params := config.Spec.Strategy.RecreateParams
 	s.retryParams = kubectl.NewRetryParams(s.retryPeriod, retryTimeout)
 	waitParams := kubectl.NewRetryParams(s.retryPeriod, retryTimeout)
 
 	if updateAcceptor == nil {
-		updateAcceptor = s.getUpdateAcceptor(time.Duration(*params.TimeoutSeconds)*time.Second, config.Spec.MinReadySeconds)
+		// updateAcceptor = s.getUpdateAcceptor(time.Duration(*params.TimeoutSeconds)*time.Second, config.Spec.MinReadySeconds)
+		updateAcceptor = s.getUpdateAcceptor(retryTimeout, config.Spec.MinReadySeconds)
```
Comment (mfojtik) [inline pkg/deploy/strategy/recreate/recreate.go]: @kargakis do we want to use the same timeout (-podStartTime) for waiting for pods to be ready? I think so.

**Pull request merged**